### DISCUSSION TITLE
Postgres AWS Deployment - RDS vs Aurora vs Aurora Serverless?

### DISCUSSION THREAD

**Discussion by lovegrovegeorge** [prisma/prisma#6989] [Q&A]
Hey all! Deploying a new project to production very soon and was interested in anyones 2 cents on selecting the database hosting as have no prior experience with Aurora or Aurora Serverless.

**Context**: The project is a community site for Cardano and their Project Catalyst funding process. Rough idea for requirements would be initially small scale users, potential for up to 100,000 users in the future and a need for availability world wide.

**Objectives (in order preference)**: Less database management, automated scaleability (initially focussed US & Europe), low price

**Deployment context** - Frontend and backend API to be all stored on Vercel as serverless functions. Choice of AWS for db hosting is due to Vercel's usage of AWS so can mirror the same regions for db deployments as their serverless functions (https://vercel.com/docs/edge-network/regions).

**Current preferences (current order preference)** - AWS Aurora, AWS Aurora Serverless, AWS RDS

**Notes**
- Aurora serverless sounds great but if always active would be more expensive and could result in being more expensive in reality - not expecting massive spikes in usage like a news site or similar.
- There isn't much information on usage of Aurora serverless from other services. Version 2 of Aurora serverless also isn't released yet so now may not be a good time to jump in.
- Aurora appears to be the easiest to manage and scale and the price difference based off the AWS calculator seems negligible for smaller to medium projects.
- RDS Proxy will be needed at some point as the API is hosted on Vercel as serverless functions meaning the database connection limit problem will occur with more traffic

**Questions**
1. Prisma + Aurora - Has anyone had any difficulty or caveats pop up with using Prisma with Aurora? I can't see any concerns and so far this is looking like the most suitable choice
2. Prisma + Aurora Serverless - Has anyone used Prisma with Aurora serverless and can share their experience (price / performance / integration)?

Thanks in advance 🍺

**Comment by janpio** [comment]
> * RDS Proxy will be needed at some point as the API is hosted on Vercel as serverless functions meaning the database connection limit problem will occur with more traffic

I have bad news for that part unfortunately: https://www.prisma.io/docs/guides/deployment/deployment-guides/caveats-when-deploying-to-aws-platforms#aws-rds-proxy :(

**Reply by lovegrovegeorge** [reply]
@janpio appreciate the response - even if you sunk my battleship! That would mean the same problem applies regardless of the AWS database then... What would be the recommended solution - setting up pgbouncer service on AWS? I also have little gage on how long I'd be able to get away without a connection pooling solution

**Reply by janpio** [reply]
That is unfortunately impossible to answer, as it totally depends on your app and traffic - also MySQL or PostgreSQL and all your configuration.

**Reply by lovegrovegeorge** [reply]
Absolutely, will have to be a wait and monitor situation. Does Prisma recommend any resources for setting up pgbouncer? That would be a great addition to that info page - 'you can't use RDS proxy because [X] but you can use pgbouncer, here are [Y] resources on integrating that'

**Reply by janpio** [reply]
Yep: https://www.prisma.io/docs/guides/performance-and-optimization/connection-management/configure-pg-bouncer
Pretty simple actually.

**Reply by lovegrovegeorge** [reply]
Perfect, I'll take a look and try and integrate something this week. Will loop back with any findings that may help others

**Reply by depombo** [reply]
If you split out the frontend and the backend, you can host the backend in [AnyCloud](https://anycloudapp.com) and you shouldn't hit the database connection limit problem since the connection will always start from the same VM as opposed to serverless offerings like Lambda or Vercel. Here is a tutorial on how to deploy a TS Node.js server that persists data to PostgreSQL via Prisma in AWS https://docs.anycloudapp.com/documentation/tutorials/aws-prisma-node-psql

Happy to talk if you want to learn more or I can help in any way!

**Reply by lovegrovegeorge** [reply]
Hey @depombo nice I like the look of what you're doing, I think the setup I have with Vercel / Next Auth JS is a pretty good setup if I can resolve the connection pooling issue, though always open to talk what you're working on, could pop me an email if you like - lovegrovegeorge@gmail.com

**Reply by depombo** [reply]
Down to connect and hear about what you are building as well! Just sent you an email, @georgelovegrove 

Got it. You could also host the isomorphic Next  + Next Auth server in AWS directly instead of with Vercel. The latency between the user and the server would increase, but the latency between the server and Postgres would decrease while alleviating the connection pooling issue. That said, I'm curious if there is another solution for the Postgres connection issue with Vercel's edge + serverless setup talking to a Postgres instance hosted in a cloud provider 🤔

**Reply by depombo** [reply]
> ..the latency between the server and Postgres would decrease while alleviating the connection pooling issue...

nvm this, just saw that you were planning to colocate the db deployment and the Vercel serverless hosting region

**Reply by lovegrovegeorge** [reply]
@janpio So update, couldn't find many good resources on setting up pgbouncer and handling that yourself in AWS so decided to play with Digital Ocean setup and was effortless. I load tested on the weekend trying Digital Ocean + Vercel both in London and the performance was more than fine for what is needed for short to mid term. Will just improve everything as I go and when needed, may move full to AWS in the future though preferably I'll decentralise parts of the stack over time instead.

I also found someone who had used Aurora Serverless, after some conversation it felt like unless you're sure you need it for the use cases just stick to the managed option for now. For the moment from what I've seen Digital Ocean is the easiest way to get setup for the minute and can scale fine for most small to medium projects.

@depombo didn't find your email in inbox at all but here is a linked to the docs for the community site i'm building - https://www.notion.so/cardanocommunity/Project-Catalyst-Community-Site-2b0616bedeb94c13a135d565eb883412

projectcatalyst.org - Literally deployed an initial version last night and will start iterating

**Reply by janpio** [reply]
That is great @georgelovegrove - absolutely valid decision, especially if you liked what you saw in your performance tests.

I created a docs issue on our side so we hopefully can improve our guidance about PgBouncer in the future: https://github.com/prisma/docs/issues/1717

**Reply by lovegrovegeorge** [reply]
Appreciate that @janpio. Everything has been flawless using everything so far but the connection pooling is definitely the least well documented for setup and maintenance in general online so could see that being the most difficult area for most newcomers coming into the stack. Vercel and their team are also absolute beasts - makes it very hard not to use their service at this point!!
